Add support to restart the model from an intermediate step

SarahAlidoost opened the pull request:
closes #9 

:red_circle: it would be better if #36 gets merged first before merging this PR. Then testing notebook can be updated too. 

Examples in testing notebook.

vanlankveldthijs commented on dgl_ptm/dgl_ptm/model/initialize_model.py:
```diff
@@ -100,42 +111,63 @@ class PovertyTrapModel(Model):
     }
 
-    def __init__(self,*, model_identifier=None, restart=False, savestate=None):
+    def __init__(self,*, model_identifier, restart=None, savestate=True):
```
We probably want to store with the saved state, the step that the process was in when the state was saved.
Then, we can have the process re-start from either a specific step (restart=X, where X is that specific step), or restart from the latest stored step (restart=-1), or start a new process (restart=None; fail on existing model storage path).

vanlankveldthijs replied:
Additionally, we still probably want to store in some way the current version of the code used to create the save state, as opposed to only the version of the model.

We may be able to do this by adding a (text) file to the program that contains a git hash that can be compared to a hash stored in the model log. This git hash file can then for example be updated on every commit, or on release.

Not sure whether differing hashes should throw a warning or an error. Also, we should probably discuss this with the researchers, because they may only really be concerned with the model version.

meiertgrootes replied:
I like the idea of linking code versions via a git hash. In general researchers will probably be mostly concerned with model version, but if you want to go for full-fledged FAIR, then code version would also be necessary. IMO a warning should suffice on mismatch.

SarahAlidoost replied:
We can also store the version of `dgl_ptm` next to the model files. I noticed that `data_collection` module also stores some data. I created issue #52 for adding software version either via version or git hash to all information stored by the model.

meiertgrootes commented on dgl_ptm/dgl_ptm/model/initialize_model.py:
```diff
@@ -315,21 +361,75 @@ def _initialize_agents_tec(self):
+            print(f'performing step {self.step_count} of {self.step_target}')
+            ptm_step(self.model_graph, self.model_data, self.step_count, self.steering_parameters)
 
+            # store each step
```
Storing the model at each step can have detrimental effects on the speed, as well as potentially being very storage intensive. It might be better to add a specification of the frequency (every x steps) with which a save state is created.

SarahAlidoost replied:
it is fixed now. It only saves every x step.

meiertgrootes commented on dgl_ptm/dgl_ptm/model/initialize_model.py:
```diff
@@ -186,6 +221,8 @@ def set_model_parameters(self,*,parameterFilePath=None, default=True, **kwargs):
                 else:
                     raise ValueError('default model has not been selected, but no model parameters have been supplied')
 
+        if self.restart:
```
Why are we setting step count to restart -1? the save state should reflect the state of the system at the end of that stepn(x), so restart should intialize to that state, and then the following step would be x+1, right?

meiertgrootes commented on dgl_ptm/dgl_ptm/model/initialize_model.py:
```diff
@@ -315,21 +361,75 @@ def _initialize_agents_tec(self):
+
+
+def _save_model(path, dictionary):
+    """ save the model_graph and generator_state in files."""
```
Is this correct that we are creating a dictionary saving the all the model states for each step, and then writing to file/updatiing the ffile version on demand? This seems problematic in terms of memory implications.

meiertgrootes replied:
Just checked. The current implementation indeeds creates a dictionary of dictionary wit all saved sates. While it may not be a problem at the number of agents and steps used in testing this will become an issue in production and should be addressed

SarahAlidoost replied:
It is fixed now. Now it only saves every X step(overwriting the previous state) and loads the last saved step. So the memory and disk space would not be a problem.

meiertgrootes commented on dgl_ptm/dgl_ptm/model/initialize_model.py:
```diff
@@ -100,42 +111,63 @@ class PovertyTrapModel(Model):
+
+        if self.restart:
+            path_model_graph = f'./{self._model_identifier}/model_graphs.bin'
+            path_generator_state = f'./{self._model_identifier}/generator_state.bin'
```
I like the idea of storing the generator states

SarahAlidoost replied:
This is needed to control the randomness after re-starting the model.

meiertgrootes reviewed (commented):
I like where this is going. I do, however, agree with some of the points thijs is raising. I also have some worries about the choice regarding saving at each timestep and what looks like maintaining te full set of saved states in memory. Maybe worth discussing in terms of approach.

SarahAlidoost commented:
@vanlankveldthijs and @meiertgrootes thank you for your comments and suggestions. I addressed them. Could you have another look at this PR and let me know if there is something else?

SarahAlidoost commented:
@meiertgrootes  I addressed your comment regarding saving a copy of the state only when it is supposed to be saved, see my [commit](https://github.com/SDCCA/DGL-PTM/pull/43/commits/158aee858e4be232f77b1b94084cc175b37cff4f).

@meiertgrootes and @vanlankveldthijs can you please approve this PR if changes are fine?

meiertgrootes reviewed (approved):
Thanks @SarahAlidoost. The changes look. good. happy to approve
